packages/opentelemetry-instrumentation-http/src/http.ts

```diff
@@ -21,144 +21,188 @@ import {
 
 /**
  * Http instrumentation plugin for Opentelemetry
  */
-export class HttpPlugin extends BasePlugin<Http> {
-  readonly component: string;
-  protected _config!: HttpPluginConfig;
+export class HttpInstrumentation extends InstrumentationBase<Http> {
   /** keep track on spans not ended */
-  private readonly _spanNotEnded: WeakSet<Span>;
-
+  private readonly _spanNotEnded: WeakSet<Span> = new WeakSet<Span>();
+  private readonly _version = process.versions.node;
   private readonly _emptySpanContext: SpanContext = {
     traceId: '',
     spanId: '',
     traceFlags: TraceFlags.NONE,
   };
 
-  constructor(readonly moduleName: string, readonly version: string) {
-    super(`@opentelemetry/plugin-${moduleName}`, VERSION);
-    // For now component is equal to moduleName but it can change in the future.
-    this.component = this.moduleName;
-    this._spanNotEnded = new WeakSet<Span>();
-    this._config = {};
-  }
-
-  /** Patches HTTP incoming and outcoming request functions. */
-  protected patch() {
-    this._logger.debug(
-      'applying patch to %s@%s',
-      this.moduleName,
-      this.version
-    );
-
-    shimmer.wrap(
-      this._moduleExports,
-      'request',
-      this._getPatchOutgoingRequestFunction()
+  constructor(config: HttpInstrumentationConfig & InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-http',
+      VERSION,
+      Object.assign({}, config)
     );
+  }
 
-    // In Node >=8, http.get calls a private request method, therefore we patch it
-    // here too.
-    if (semver.satisfies(this.version, '>=8.0.0')) {
-      shimmer.wrap(
-        this._moduleExports,
-        'get',
-        this._getPatchOutgoingGetFunction(this._moduleExports.request)
-      );
-    }
-
-    if (
-      this._moduleExports &&
-      this._moduleExports.Server &&
-      this._moduleExports.Server.prototype
-    ) {
-      shimmer.wrap(
-        this._moduleExports.Server.prototype,
-        'emit',
-        this._getPatchIncomingRequestFunction()
-      );
-    } else {
-      this._logger.error(
-        'Could not apply patch to %s.emit. Interface is not as expected.',
-        this.moduleName
-      );
-    }
+  private _getConfig(): HttpInstrumentationConfig {
+    return this._config;
+  }
 
-    return this._moduleExports;
+  setConfig(config: HttpInstrumentationConfig & InstrumentationConfig = {}) {
+    this._config = Object.assign({}, config);
   }
 
-  /** Unpatches all HTTP patched function. */
-  protected unpatch(): void {
-    shimmer.unwrap(this._moduleExports, 'request');
-    if (semver.satisfies(this.version, '>=8.0.0')) {
-      shimmer.unwrap(this._moduleExports, 'get');
-    }
-    if (
-      this._moduleExports &&
-      this._moduleExports.Server &&
-      this._moduleExports.Server.prototype
-    ) {
-      shimmer.unwrap(this._moduleExports.Server.prototype, 'emit');
-    }
+  init() {
+    const httpModule = new InstrumentationNodeModuleDefinition<Http>(
```

Comment:
I would just split it into 2 private functions for example `patchHttp` `patchHttps`